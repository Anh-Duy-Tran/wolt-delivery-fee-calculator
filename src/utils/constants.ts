// Constants for ruleId: FREE_DELIVERY_CART_VALUE_BASED
export const THRESHOLD_FREE_CART_VALUE = 200; // euros

// Constants for ruleId: SURCHARGE_SMALL_ORDER
export const SMALL_ORDER_SURCHARGE_THRESHOLD = 10; // euros

// Constants for ruleId: SURCHARGE_DELIVERY_DISTANCE
export const BASE_DISTANCE_LIMIT = 1000; // meter
export const BASE_DELIVERY_FEE = 2; // euros
export const ADDITIONAL_DISTANCE_UNIT = 500; // meter
export const ADDITIONAL_DISTANCE_FEE = 1; // euros

// Constants for ruleId: SURCHARGE_ITEM_COUNT
export const ITEM_SURCHARGE_THRESHOLD = 5; // number of items
export const ITEM_SURCHARGE_AMOUNT = 0.5; // euros
export const BULK_ITEM_THRESHOLD = 12; // number of items
export const BULK_SURCHARGE_AMOUNT = 1.2; // euros

// Constants for ruleId: RUSH_HOUR
export const RUSH_HOUR_DAY_OF_WEEK = 5; // Friday
export const RUSH_HOUR_START_TIME = 15; // 15:00 or 3:00 PM
export const RUSH_HOUR_END_TIME = 19; // 19:00 or 7:00 PM
export const RUSH_HOUR_MULTIPLIER = 1.2; // x1.2

// Constants for ruleId: MAX_FEE
export const MAX_DELIVERY_FEE = 15; // euros
