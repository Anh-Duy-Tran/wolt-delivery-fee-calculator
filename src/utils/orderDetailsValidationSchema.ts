import * as Yup from 'yup';

export const validationSchema = Yup.object().shape({
  cartValue: Yup.number()
    .transform((value) => (isNaN(value) ? undefined : value))
    .required('Cart value is required.')
    .positive('Cart value must be a positive number.'),
  deliveryDistance: Yup.number()
    .transform((value) => (isNaN(value) ? undefined : value))
    .required('Delivery distance is required.')
    .positive('Delivery distance must be a positive number.')
    .integer('Delivery distance must be a whole number.'),
  numberOfItems: Yup.number()
    .transform((value) => (isNaN(value) ? undefined : value))
    .required('Number of items is required.')
    .positive('Number of items must be a positive number.')
    .integer('Number of items must be a whole number.'),
  orderTime: Yup.date()
    .transform(function (value) {
      return this.isType(value) ? value : undefined;
    })
    .required('Please enter a valid time.'),
});
