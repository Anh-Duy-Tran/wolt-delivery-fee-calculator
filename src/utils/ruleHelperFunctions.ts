/**
 * The delivery is free when the cart value is equal or more than 200€.
 * @param cartValue Value of the shopping cart in euros.
 * @return boolean represent the eligibility of the order to free delivery.
 */
export function eligibleForFreeDeliveryBasedOnCartValue(
  cartValue: number
): boolean {
  const threshold = 200; //Euro
  return cartValue >= threshold;
}

/**
 * If the cart value is less than 10€, a small order surcharge is added to the
 * delivery price. The surcharge is the difference between the cart value and
 * 10€.
 * For example if the cart value is 8.90€, the surcharge will be 1.10€.
 * @param cartValue Value of the shopping cart in euros.
 * @return surcharge amount in euros.
 */
export function calculateSmallOrderSurcharge(cartValue: number): number {
  const smallOrderValueThreshold = 10; // Euro

  if (cartValue < smallOrderValueThreshold) {
    return Math.max(smallOrderValueThreshold - cartValue, 0);
  }

  return 0;
}

/**
 * A delivery fee for the first 1000 meters (=1km) is 2€. If the delivery
 * distance is longer than that, 1€ is added for every additional 500 meters
 * that the courier needs to travel before reaching the destination. Even if the
 * distance would be shorter than 500 meters, the minimum fee is always 1€.
 * - Example 1: If the delivery distance is 1499 meters, the delivery fee is: 2€
 *   base fee + 1€ for the additional 500 m => 3€
 * - Example 2: If the delivery distance is 1500 meters, the delivery fee is: 2€
 *   base fee + 1€ for the additional 500 m => 3€
 * - Example 3: If the delivery distance is 1501 meters, the delivery fee is: 2€
 *   base fee + 1€ for the first 500 m + 1€ for the second 500 m => 4€
 * @param distance The distance between the store and customer’s location in meters.
 * @returns surcharge amount in euros (cannot be 0 in this case => rule always apply).
 */
export function calculateDeliveryDistanceSurcharge(distance: number): number {
  const fixPriceDistance = 1000; // Meter
  const fixPrice = 2; // Euro
  const distanceInterval = 500; // Meter
  const pricePerInterval = 1; // Euro

  return (
    fixPrice +
    Math.ceil(Math.max(distance - fixPriceDistance, 0) / distanceInterval) *
      pricePerInterval
  );
}

/**
 *If the number of items is five or more, an additional 50 cent surcharge is
 *added for each item above and including the fifth item. An extra "bulk" fee
 *applies for more than 12 items of 1,20€
 * - Example 1: If the number of items is 4, no extra surcharge
 * - Example 2: If the number of items is 5, 50 cents surcharge is added
 * - Example 3: If the number of items is 10, 3€ surcharge (6 x 50 cents) is
 *   added
 * - Example 4: If the number of items is 13, 5,70€ surcharge is added ((9 * 50
 *   cents) + 1,20€)
 * - Example 5: If the number of items is 14, 6,20€ surcharge is added ((10 * 50
 *   cents) + 1,20€)
 * @param numberOfItems The number of items in the customer's shopping cart.
 * @returns surcharge amount in euros.
 */
export function calculateNumberOfItemSurcharge(numberOfItems: number): number {
  const threshold = 4; // Number of items without surcharge
  const bulkThreshold = 12; // Number of Items
  const pricePerItem = 0.5; // Euro
  const bulkPrice = 1.2; // Euro

  let surcharge = 0;

  if (numberOfItems > threshold) {
    surcharge += (numberOfItems - threshold) * pricePerItem;
  }
  // Adding bulk price if the number of item surpass the bulk number of item threshold
  if (numberOfItems > bulkThreshold) {
    surcharge += bulkPrice;
  }

  return surcharge;
}

/**
 * During the Friday rush, 3 - 7 PM, the delivery fee (the total fee including
 * possible surcharges) will be multiplied by 1.2x.
 * @param orderTime The date/time when the order is being made.
 * @param currentFee The current calculated delivery fee for the order in euros.
 * @returns surcharge amount in euros.
 */
export function calculateRushHourSurcharge(
  orderTime: Date,
  currentFee: number
): number {
  const rushDay = 5; // Friday
  const lowerRushHour = 15; // 3PM or 15:00
  const higherRushHour = 19; // 7PM or 19:00
  const increaseFactor = 0.2; // 1.2 - 1 = 0.2 or 20% increase from the fee

  if (
    orderTime.getDay() === rushDay &&
    orderTime.getHours() >= lowerRushHour &&
    orderTime.getHours() < higherRushHour
  ) {
    return currentFee * increaseFactor;
  }

  return 0;
}
