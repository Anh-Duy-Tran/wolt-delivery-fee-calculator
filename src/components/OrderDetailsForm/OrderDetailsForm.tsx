import { InputField } from '../InputField';
import { yupResolver } from '@hookform/resolvers/yup';
import { SubmitHandler, useForm } from 'react-hook-form';
import { Button } from '../Button';
import {
  FeeCalculatorReturnType,
  deliveryFeeCaculator,
} from '../../utils/deliveryFeeCaculator';
import { Card } from '../Card';
import { validationSchema } from '../../utils/orderDetailsValidationSchema';

interface OrderDetailsFormProps {
  saveCalculatedOrderDetails: (orderDetails: OrderDetailFormType) => void;
  handleResult: (result: FeeCalculatorReturnType) => void;
}

export type OrderDetailFormType = {
  cartValue: number;
  deliveryDistance: number;
  numberOfItems: number;
  orderTime: Date;
};

export function OrderDetailsForm({
  saveCalculatedOrderDetails,
  handleResult,
}: OrderDetailsFormProps) {
  const {
    handleSubmit,
    register,
    formState: { errors },
  } = useForm<OrderDetailFormType>({
    mode: 'all',
    resolver: yupResolver(validationSchema),
    reValidateMode: 'onChange',
  });

  const onSubmit: SubmitHandler<OrderDetailFormType> = (data) => {
    handleResult(deliveryFeeCaculator(data));
    saveCalculatedOrderDetails(data);
  };

  return (
    <Card className=" flex flex-col gap-4">
      <h2>ORDER INFORMATION</h2>

      <form
        noValidate
        className="flex flex-col gap-2"
        onSubmit={handleSubmit(onSubmit)}
      >
        <InputField
          {...register('cartValue')}
          errorMessage={errors.cartValue?.message}
          data-testid="cartValue"
          label="Cart Value (EUR)"
          type="number"
          step={0.1}
        />
        <InputField
          {...register('deliveryDistance')}
          errorMessage={errors.deliveryDistance?.message}
          data-testid="deliveryDistance"
          label="Delivery Distance (Meter)"
          type="number"
          step={10}
        />
        <InputField
          {...register('numberOfItems')}
          errorMessage={errors.numberOfItems?.message}
          data-testid="numberOfItems"
          label="Number of Items"
          type="number"
        />
        <InputField
          {...register('orderTime')}
          errorMessage={errors.orderTime?.message}
          data-testid="orderTime"
          label="Time"
          type="datetime-local"
          placeholder="DD.MM.YYYY"
        />
        <Button data-testid={'calculateFee'} className="mt-2" type="submit">
          CALCULATE FEE
        </Button>
      </form>
    </Card>
  );
}
