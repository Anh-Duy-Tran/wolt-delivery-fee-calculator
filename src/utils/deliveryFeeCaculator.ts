import { OrderDetailFormType } from '../components/OrderDetailsForm/OrderDetailsForm';
import { validationSchema } from './orderDetailsValidationSchema';

export const MAX_DELIVERY_FEE = 15; // Euro

type AdditionalFeeRuleType = {
  type: string;
  message: string;
  calculateAdditionalFee: (orderDetails: OrderDetailFormType) => number;
};

type FreeDeliveryRuleType = {
  type: string;
  message: string;
  condition: (orderDetails: OrderDetailFormType) => boolean;
};

type FeeBasedRuleType = {
  type: string;
  message: string;
  calculateFeeBasedAdditionalFee: (
    orderDetails: OrderDetailFormType,
    currentFee: number
  ) => number;
};

export type FeeCalculatorReturnType = {
  deliveryFee: number;
  subjectedRules: {
    type: string;
    message: string;
    amount: string;
  }[];
};

export const additionalFeeRules: AdditionalFeeRuleType[] = [
  {
    type: 'smallOrder',
    message:
      'For orders under 10€, a small order surcharge equal to the difference up to 10€ will be added to your delivery fee.',
    calculateAdditionalFee: ({ cartValue }) => {
      const smallOrderThreshold = 10; // Euro

      return Math.max(smallOrderThreshold - cartValue, 0);
    },
  },
  {
    type: 'deliveryDistance',
    message:
      'Delivery fee: 2€ for the first 1km, plus 1€ for every additional 500m (minimum 1€ surcharge for distances over 1km).',
    calculateAdditionalFee: ({ deliveryDistance }) => {
      const fixPriceDistance = 1000; // Meter
      const fixPrice = 2; // Euro
      const distanceInterval = 500; // Meter
      const pricePerInterval = 1; // Euro

      return (
        fixPrice +
        Math.ceil(
          Math.max(deliveryDistance - fixPriceDistance, 0) / distanceInterval
        ) *
          pricePerInterval
      );
    },
  },
  {
    type: 'itemCount',
    message:
      'Orders with 5+ items incur a 50 cent surcharge per item, starting from the 5th item, and a 1,20€ bulk fee for more than 12 items.',
    calculateAdditionalFee: ({ numberOfItems }) => {
      const threshold = 4; // Number of Items
      const bulkThreshold = 12; // Number of Items
      const pricePerItem = 0.5; // Euro
      const bulkPrice = 1.2; // Euro

      let surcharge = 0; // Euro

      if (numberOfItems > threshold) {
        surcharge += (numberOfItems - threshold) * pricePerItem;
      }

      if (numberOfItems > bulkThreshold) {
        surcharge += bulkPrice;
      }

      return surcharge;
    },
  },
];

export const freeDeliveryRules: FreeDeliveryRuleType[] = [
  {
    type: 'freeDelivery',
    message: 'Enjoy free delivery for order of 200€ or more.',
    condition: ({ cartValue }) => {
      const threshold = 200; //Euro

      return cartValue >= threshold;
    },
  },
];

export const feeBasedRules: FeeBasedRuleType[] = [
  {
    type: 'rushHour',
    message:
      'During Friday rush hours (3 - 7 PM Local Time Zone), delivery fees are increased by 1.2x.',
    calculateFeeBasedAdditionalFee: (orderDetails, currentFee) => {
      const rushDay = 5; // Friday
      const lowerRushHour = 15; // 3PM or 15:00
      const higherRushHour = 19; // 7PM or 19:00
      const increaseFactor = 0.2; // 1.2 - 1 = 0.2 or 20% increase from the fee

      const { orderTime } = orderDetails;
      if (
        orderTime.getDay() === rushDay &&
        orderTime.getHours() >= lowerRushHour &&
        orderTime.getHours() < higherRushHour
      ) {
        return currentFee * increaseFactor;
      }

      return 0;
    },
  },
];

export function deliveryFeeCaculator(
  orderDetails: OrderDetailFormType
): FeeCalculatorReturnType {
  if (!validationSchema.isValidSync(orderDetails)) {
    throw new TypeError('Invalid Order Details');
  }

  const deliveryFeeResponse: FeeCalculatorReturnType = {
    deliveryFee: 0,
    subjectedRules: [],
  };

  for (const rule of freeDeliveryRules) {
    const { type, message, condition } = rule;
    if (condition(orderDetails)) {
      deliveryFeeResponse.subjectedRules.push({
        type,
        message,
        amount: 'FREE',
      });

      // Stop the calculation if any match of the free delivery rules
      return deliveryFeeResponse;
    }
  }

  for (const rule of additionalFeeRules) {
    const { type, message, calculateAdditionalFee } = rule;
    // Get the possible surcharge based only on the order details
    const fee = calculateAdditionalFee(orderDetails);
    if (fee) {
      deliveryFeeResponse.deliveryFee += fee;
      deliveryFeeResponse.subjectedRules.push({
        type,
        message,
        amount: `${fee.toFixed(2)}€`,
      });
    }
  }

  for (const rule of feeBasedRules) {
    const { type, message, calculateFeeBasedAdditionalFee } = rule;
    // Get the possible surcharge based on the order details and the current calculated fee
    const fee = calculateFeeBasedAdditionalFee(
      orderDetails,
      deliveryFeeResponse.deliveryFee
    );
    if (fee) {
      deliveryFeeResponse.deliveryFee += fee;
      deliveryFeeResponse.subjectedRules.push({
        type,
        message,
        amount: `${fee.toFixed(2)}€`,
      });
    }
  }

  // Maximum fee rule
  if (deliveryFeeResponse.deliveryFee > MAX_DELIVERY_FEE) {
    deliveryFeeResponse.deliveryFee = MAX_DELIVERY_FEE;
    deliveryFeeResponse.subjectedRules.push({
      type: 'maxFee',
      message: `Maximum delivery fee of ${MAX_DELIVERY_FEE}€ reached`,
      amount: '',
    });
  }

  return deliveryFeeResponse;
}
