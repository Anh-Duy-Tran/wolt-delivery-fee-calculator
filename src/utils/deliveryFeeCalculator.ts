import { OrderDetailFormType } from '../components/OrderDetailsForm/OrderDetailsForm';
import {
  calculateDeliveryDistanceSurcharge,
  calculateNumberOfItemSurcharge,
  calculateRushHourSurcharge,
  calculateSmallOrderSurcharge,
  eligibleForFreeDeliveryBasedOnCartValue,
} from './ruleHelperFunctions';
import { validationSchema } from './orderDetailsValidationSchema';
import { MAX_DELIVERY_FEE } from './constants';

export enum DeliveryFeeRuleId {
  // Free delivery rules
  FreeDeliveryCartValueBased = 'FREE_DELIVERY_CART_VALUE_BASED',

  // Surcharge rule
  SurchargeSmallOrder = 'SURCHARGE_SMALL_ORDER',
  SurchargeDeliveryDistance = 'SURCHARGE_DELIVERY_DISTANCE',
  SurchargeItemCount = 'SURCHARGE_ITEM_COUNT',
  SurchargeRushHour = 'RUSH_HOUR',

  // Other rule
  MaxFee = 'MAX_FEE',
}

type FeeAdjustmentResult =
  | { isRuleApplied: true; newFee: number; feeChangeAmount?: number }
  | { isRuleApplied: false };

type DeliveryFeeRuleType = {
  /**
   * A short message describe the rule to the client
   */
  message: string;
  /**
   * Set to True if the chain of responsibilities should end if this particular rule is applied
   */
  shouldTerminateCalculationIfApplied?: boolean;
  /**
   * Function try and apply rule to the requested order details and return the fee adjustment instructions.
   * @param requestedOrderDetails the requested order details object
   * @param currentFee the current calculated delivery fee
   * @returns fee adjustment result object (type FeeAdjustmentResult)
   */
  getFeeAdjustment: (
    requestedOrderDetails: OrderDetailFormType,
    currentFee: number
  ) => FeeAdjustmentResult;
};

export type FeeCalculatorReturnType = {
  /**
   * The calculated delivery fee
   */
  deliveryFee: number;
  /**
   * Array of object that stores the information about the applied rules
   */
  subjectedRules: {
    id: DeliveryFeeRuleId;
    message: string;
    amount: string;
  }[];
};

export type DeliveryFeeRulesMapType = {
  [key in DeliveryFeeRuleId]: DeliveryFeeRuleType;
};

const RULE_NOT_APPLY: FeeAdjustmentResult = {
  isRuleApplied: false,
};

export const deliveryFeeRulesMap: DeliveryFeeRulesMapType = {
  [DeliveryFeeRuleId.FreeDeliveryCartValueBased]: {
    message: 'Enjoy free delivery for order of 200€ or more.',
    shouldTerminateCalculationIfApplied: true,
    getFeeAdjustment: ({ cartValue }) =>
      eligibleForFreeDeliveryBasedOnCartValue(cartValue)
        ? {
            isRuleApplied: true,
            newFee: 0,
          }
        : RULE_NOT_APPLY,
  },
  [DeliveryFeeRuleId.SurchargeSmallOrder]: {
    message:
      'For orders under 10€, a small order surcharge equal to the difference up to 10€ will be added to your delivery fee.',
    getFeeAdjustment: ({ cartValue }, currentFee) => {
      const surcharge = calculateSmallOrderSurcharge(cartValue);
      return surcharge !== 0
        ? {
            isRuleApplied: true,
            newFee: currentFee + surcharge,
            feeChangeAmount: surcharge,
          }
        : RULE_NOT_APPLY;
    },
  },
  [DeliveryFeeRuleId.SurchargeDeliveryDistance]: {
    message:
      'Delivery fee: 2€ for the first 1km, plus 1€ for every additional 500m (minimum 1€ surcharge for distances over 1km).',
    getFeeAdjustment: ({ deliveryDistance }, currentFee) => {
      const surcharge = calculateDeliveryDistanceSurcharge(deliveryDistance);
      return surcharge !== 0
        ? {
            isRuleApplied: true,
            newFee: currentFee + surcharge,
            feeChangeAmount: surcharge,
          }
        : RULE_NOT_APPLY;
    },
  },
  [DeliveryFeeRuleId.SurchargeItemCount]: {
    message:
      'Orders with 5+ items incur a 50 cent surcharge per item, starting from the 5th item, and a 1,20€ bulk fee for more than 12 items.',
    getFeeAdjustment: ({ numberOfItems }, currentFee) => {
      const surcharge = calculateNumberOfItemSurcharge(numberOfItems);
      return surcharge !== 0
        ? {
            isRuleApplied: true,
            newFee: currentFee + surcharge,
            feeChangeAmount: surcharge,
          }
        : RULE_NOT_APPLY;
    },
  },
  [DeliveryFeeRuleId.SurchargeRushHour]: {
    message:
      'During Friday rush hours (3 - 7 PM Local Time Zone), delivery fees are increased by 1.2x.',
    getFeeAdjustment: ({ orderTime }, currentFee) => {
      const surcharge = calculateRushHourSurcharge(orderTime, currentFee);
      return surcharge !== 0
        ? {
            isRuleApplied: true,
            newFee: currentFee + surcharge,
            feeChangeAmount: surcharge,
          }
        : RULE_NOT_APPLY;
    },
  },
  [DeliveryFeeRuleId.MaxFee]: {
    message: `Maximum delivery fee of 15€ reached`,
    getFeeAdjustment: (_, currentFee) =>
      currentFee > MAX_DELIVERY_FEE
        ? {
            isRuleApplied: true,
            newFee: MAX_DELIVERY_FEE,
          }
        : RULE_NOT_APPLY,
  },
};

const deliveryFeeRuleIdsInOrder: DeliveryFeeRuleId[] = [
  // Rules that would terminate the calculation if applied should be run first
  // Example: Free delivery Rule should run first
  DeliveryFeeRuleId.FreeDeliveryCartValueBased,

  // Normal surcharge Rules
  DeliveryFeeRuleId.SurchargeSmallOrder,
  DeliveryFeeRuleId.SurchargeDeliveryDistance,
  DeliveryFeeRuleId.SurchargeItemCount,

  // Surcharge multiplier Rules should be run after all but before the maximum fee limit Rules
  DeliveryFeeRuleId.SurchargeRushHour,

  // Maximum fee limit Rules
  DeliveryFeeRuleId.MaxFee,
];

/**
 * Calculates the delivery fee based on various rules.
 * - Small order surcharge is added for orders less than 10€.
 * - Delivery distance fee: 2€ for the first 1km, then 1€ for every additional 500m.
 * - Item count surcharge: 50 cents per item for 5+ items; additional bulk fee for 12+ items.
 * - Max fee cap at 15€.
 * - Free delivery for orders 200€ and above.
 * - Rush hour (Fri, 3-7 PM Local Time) fee increase by 1.2x.
 *
 * @param orderDetails The details of the order.
 * @returns The calculated fee and applied rules.
 */
export function deliveryFeeCalculator(
  orderDetails: OrderDetailFormType
): FeeCalculatorReturnType {
  // Validate the order details request object
  if (!validationSchema.isValidSync(orderDetails)) {
    throw new TypeError('Invalid Order Details');
  }

  let deliveryFeeResponse: FeeCalculatorReturnType = {
    deliveryFee: 0,
    subjectedRules: [],
  };

  // Loop through all the rules in an ordered matter
  for (const ruleId of deliveryFeeRuleIdsInOrder) {
    const rule = deliveryFeeRulesMap[ruleId];
    const currentFee = deliveryFeeResponse.deliveryFee;

    const { message, getFeeAdjustment } = rule;
    const feeAdjustment = getFeeAdjustment(orderDetails, currentFee);

    if (feeAdjustment.isRuleApplied) {
      // Modify response by construct new response object (the functional
      // programming way, the alternative would be modifying the object in-place
      // would technically be more efficient though it does not adhere to
      // functional programming paradigms)
      deliveryFeeResponse = {
        deliveryFee: feeAdjustment.newFee,
        subjectedRules: [
          ...deliveryFeeResponse.subjectedRules,
          {
            id: ruleId,
            message,
            amount: feeAdjustment.feeChangeAmount
              ? `${feeAdjustment.feeChangeAmount.toFixed(2)}€`
              : '',
          },
        ],
      };

      if (rule.shouldTerminateCalculationIfApplied) {
        // Terminate the calculation early, return the current response
        return deliveryFeeResponse;
      }
    }
  }

  return deliveryFeeResponse;
}
