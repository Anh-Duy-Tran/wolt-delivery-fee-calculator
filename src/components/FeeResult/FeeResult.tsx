import { FeeCalculatorReturnType } from '../../utils/deliveryFeeCaculator';
import { Card } from '../Card';
import { InfoTooltip } from '../InfoTooltip';
import { OrderDetailFormType } from '../OrderDetailsForm/OrderDetailsForm';

interface FeeResultProps {
  orderDetails?: OrderDetailFormType;
  result?: FeeCalculatorReturnType;
}

export function FeeResult({ orderDetails, result }: FeeResultProps) {
  return (
    <Card className="flex flex-col gap-4">
      <h2>CALCULATION RESULT</h2>
      {result && orderDetails ? (
        <>
          <div>
            <h3>Order details:</h3>
            <div className="indented grid grid-cols-2">
              <p data-testid={'submittedCartValue'}>
                {`• Cart Value: ${orderDetails.cartValue.toFixed(2)}€`}
              </p>
              <p data-testid={'submittedDeliveryDistance'}>
                {`• Distance: ${orderDetails.deliveryDistance}m`}
              </p>
              <p data-testid={'submittedNumberOfItems'}>
                {`• Number of items: ${orderDetails.numberOfItems}`}
              </p>
              <p data-testid={'submittedOrderTime'}>
                {`• Time: ${orderDetails.orderTime.toDateString()} at ${orderDetails.orderTime.toLocaleTimeString()}`}
              </p>
            </div>
          </div>
          <div className="relative">
            <h3>Delivery price:</h3>
            <h1
              data-testid={'fee'}
              className="indented bg-gradient-to-r from-primary to-secondary bg-clip-text text-transparent"
            >
              {`${result.deliveryFee.toFixed(2)}€`}
            </h1>
            <InfoTooltip
              data-testid={'info-icon'}
              tipContent={
                <ul className="flex list-disc flex-col gap-2">
                  {result.subjectedRules.map((rule, i) => (
                    <li key={i}>{`${rule.message} ( ${rule.amount} )`}</li>
                  ))}
                </ul>
              }
            >
              {infoIcon}
            </InfoTooltip>
          </div>
        </>
      ) : (
        <div className="w-full text-center italic text-black/50">
          Fill in order details and press calculate to get result.
        </div>
      )}
    </Card>
  );
}

const infoIcon = (
  <svg
    width="24"
    height="24"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path d="M11.5 16.8v-1.2h1v1.2h-1zm0-9.6v7.2h1V7.2h-1z"></path>
    <path
      fillRule="evenodd"
      clipRule="evenodd"
      d="M12 21.6a9.6 9.6 0 0 0 9.6-9.6 9.6 9.6 0 1 0-19.2 0 9.6 9.6 0 0 0 9.6 9.6zm0-1a8.6 8.6 0 1 0 0-17.2 8.6 8.6 0 0 0 0 17.2z"
    ></path>
  </svg>
);
