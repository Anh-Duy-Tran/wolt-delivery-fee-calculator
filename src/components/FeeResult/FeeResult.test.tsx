import { render, screen, fireEvent } from '@testing-library/react';
import { expect, it, describe } from 'vitest';
import { DEFAULT_MESSAGE, FeeResult } from './FeeResult';
import { OrderDetailFormType } from '../OrderDetailsForm/OrderDetailsForm';
import { FeeCalculatorReturnType } from '../../utils/deliveryFeeCaculator';

const mockOrderDetails: OrderDetailFormType = {
  cartValue: 100,
  deliveryDistance: 2000,
  numberOfItems: 3,
  orderTime: new Date('2024-01-21T12:00:00'),
};

const mockResult: FeeCalculatorReturnType = {
  deliveryFee: 15,
  subjectedRules: [
    { type: 'baseFee', message: 'Base fee', amount: '5€' },
    { type: 'distanceFee', message: 'Distance fee', amount: '10€' },
  ],
};

describe('FeeResult Component', () => {
  // Snapshot test
  it('renders correctly', () => {
    const { asFragment } = render(<FeeResult />);
    expect(asFragment()).toMatchSnapshot();
  });

  // Snapshot test
  it('renders correctly with props', () => {
    const { asFragment } = render(
      <FeeResult orderDetails={mockOrderDetails} result={mockResult} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  // Test for rendering with no props
  it('renders with default message when no props are provided', () => {
    render(<FeeResult />);
    expect(screen.getByText(DEFAULT_MESSAGE)).toBeInTheDocument();
  });

  // Test for rendering with props
  it('renders order details and fee result when props are provided', () => {
    render(<FeeResult orderDetails={mockOrderDetails} result={mockResult} />);
    expect(screen.getByTestId('submittedCartValue')).toHaveTextContent(
      '• Cart Value: 100.00€'
    );
    expect(screen.getByTestId('submittedDeliveryDistance')).toHaveTextContent(
      '• Distance: 2000m'
    );
    expect(screen.getByTestId('submittedNumberOfItems')).toHaveTextContent(
      '• Number of items: 3'
    );
    expect(screen.getByTestId('submittedOrderTime')).toHaveTextContent(
      `• Time: ${mockOrderDetails.orderTime.toDateString()} at ${mockOrderDetails.orderTime.toLocaleTimeString()}`
    );
    expect(screen.getByTestId('fee')).toHaveTextContent('15.00€');
  });

  // Test for tooltip content in FeeResult
  it('passes correct tooltip content to InfoTooltip', () => {
    render(<FeeResult orderDetails={mockOrderDetails} result={mockResult} />);

    const infoIcon = screen.getByTestId('info-icon');
    fireEvent.mouseOver(infoIcon);

    // Check if the content passed to InfoTooltip is rendered
    mockResult.subjectedRules.forEach((rule) => {
      expect(
        screen.getByText(`${rule.message} ( ${rule.amount} )`)
      ).toBeInTheDocument();
    });
  });

  it('displays tooltip with fee calculation rules on hover', () => {
    render(<FeeResult orderDetails={mockOrderDetails} result={mockResult} />);
    const infoIcon = screen.getByTestId('info-icon');
    fireEvent.mouseOver(infoIcon);

    const tooltip = screen.getByTestId('tooltip');
    expect(tooltip).toBeInTheDocument();
    expect(tooltip).toBeVisible();
  });

  it('displays tooltip with fee calculation rules on focus', () => {
    render(<FeeResult orderDetails={mockOrderDetails} result={mockResult} />);
    const infoIcon = screen.getByTestId('info-icon');
    fireEvent.focus(infoIcon);

    const tooltip = screen.getByTestId('tooltip');
    expect(tooltip).toBeInTheDocument();
  });
});
