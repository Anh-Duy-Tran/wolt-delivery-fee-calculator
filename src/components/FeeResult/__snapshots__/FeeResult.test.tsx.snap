// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FeeResult Component > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4 w-full rounded-xl bg-white p-5 shadow-xl"
    data-testid="card"
  >
    <h2>
      CALCULATION RESULT
    </h2>
    <div
      class="w-full text-center italic text-black/50"
    >
      Fill in order details and press calculate to get result.
    </div>
  </div>
</DocumentFragment>
`;

exports[`FeeResult Component > renders correctly with props 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4 w-full rounded-xl bg-white p-5 shadow-xl"
    data-testid="card"
  >
    <h2>
      CALCULATION RESULT
    </h2>
    <div>
      <h3>
        Order details:
      </h3>
      <div
        class="indented grid grid-cols-2"
      >
        <p
          data-testid="submittedCartValue"
        >
          • Cart Value: 100.00€
        </p>
        <p
          data-testid="submittedDeliveryDistance"
        >
          • Distance: 2000m
        </p>
        <p
          data-testid="submittedNumberOfItems"
        >
          • Number of items: 3
        </p>
        <p
          data-testid="submittedOrderTime"
        >
          • Time: Sun Jan 21 2024 at 12:00:00
        </p>
      </div>
    </div>
    <div
      class="relative"
    >
      <h3>
        Delivery price:
      </h3>
      <div
        class="flex items-end justify-between"
      >
        <h1
          class="indented bg-gradient-to-br from-primary to-secondary bg-clip-text text-transparent"
          data-testid="fee"
        >
          15.00€
        </h1>
        <div
          class="relative"
        >
          <div
            class="peer rounded ring-wolt-blue focus:ring-2"
            data-testid="info-icon"
            tabindex="0"
          >
            <svg
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.5 16.8v-1.2h1v1.2h-1zm0-9.6v7.2h1V7.2h-1z"
              />
              <path
                clip-rule="evenodd"
                d="M12 21.6a9.6 9.6 0 0 0 9.6-9.6 9.6 9.6 0 1 0-19.2 0 9.6 9.6 0 0 0 9.6 9.6zm0-1a8.6 8.6 0 1 0 0-17.2 8.6 8.6 0 0 0 0 17.2z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          <div
            class="absolute z-50 right-0 mt-2 w-[350px] scale-75 rounded-md bg-slate-700 p-3 text-sm text-white opacity-0 transition peer-hover:scale-100 peer-hover:opacity-100 peer-focus:scale-100 peer-focus:opacity-100"
            data-testid="tooltip"
          >
            <ul
              class="flex list-inside list-disc flex-col gap-2"
            >
              <li>
                Base fee ( 5€ )
              </li>
              <li>
                Distance fee ( 10€ )
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
